### PR TITLE
[CELEBORN-1771] Bring forward PUSH_DATA_CREATE_CONNECTION_FAIL_REPLICA

### PR DESCRIPTION
**Pull request by zaynt4606**
### What changes were proposed in this pull request?
1.Move the peerWorker available judgement out of ThreadPool.
2.Move `retain` after the available worker judgment Which means we don't have to release if peerWorker is unavailable.
2. Add `fileWriter.decrementPendingWrites()` if peerWorker is unavailable since it will return and won't decrementPendingWreites in `writeLocalData`.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UT & cluster testing.

### CODE REVIEW

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 1]
```diff
@@ -719,7 +720,7 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
               logError(
                 s"PushMergedData replication failed during connecting peer for partitionLocation: $location",
                 e)
-              pushMergedDataCallback.onFailure(
+              wrappedCallback.onFailure(
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:723]: should keep pushMergedDataCallback

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 2]
```diff
@@ -374,7 +375,7 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
               logError(
                 s"PushData replication failed during connecting peer for partitionLocation: $location",
                 e)
-              callbackWithTimer.onFailure(
+              wrappedCallback.onFailure(
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:378]: ditto

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala (hunk 1 above)
Comment (zaynt4606) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala]: Done~

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 3]
```diff
@@ -273,26 +273,26 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
...
+          s"PushData replication failed caused by unavailable peer for partitionLocation: $location")
+        callbackWithTimer.onFailure(
+          new CelebornIOException(StatusCode.PUSH_DATA_CREATE_CONNECTION_FAIL_REPLICA))
+        return
+      }
+
       pushData.body().retain()
       replicateThreadPool.submit(new Runnable {
         override def run(): Unit = {
-          val peer = location.getPeer
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:279]: Should also check unavailablePeers when run replicate task

Comment (zaynt4606) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:279]: Done~

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 4]
```diff
@@ -582,27 +603,32 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
...
+      val peer = location.getPeer
+      val peerWorker = new WorkerInfo(
+        peer.getHost,
+        peer.getRpcPort,
+        peer.getPushPort,
+        peer.getFetchPort,
+        peer.getReplicatePort)
+      if (unavailablePeers.containsKey(peerWorker)) {
+        for (fileWriterIndex <- 0 until totalFileWriters) {
+          if (fileWriters(fileWriterIndex) != null &&
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:616]: fileWriters(fileWriterIndex) -> fileWriter

**Review by RexXiong — APPROVED** [review]
LGTM, except a nit

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala (hunk 4 above)
Comment (zaynt4606) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala]: Done~

**Review by onebox-li — APPROVED** [review]

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 5]
```diff
@@ -582,27 +603,32 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
...
+        peer.getPushPort,
+        peer.getFetchPort,
+        peer.getReplicatePort)
+      if (unavailablePeers.containsKey(peerWorker)) {
+        for (fileWriterIndex <- 0 until totalFileWriters) {
+          val fileWriter = fileWriters(fileWriterIndex)
+          if (fileWriter != null && !pushMergedDataCallback.isHardSplitPartition(fileWriterIndex)) {
+            fileWriter.decrementPendingWrites()
+          }
+        }
```
Comment (mridulm) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:620]: QQ: I am trying to understand why this is different here vs within the `Runnable` below ?

**Review by RexXiong — APPROVED** [review]
LGTM

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala (hunk 5 above)
Comment (zaynt4606) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:620]: To reduce the number of jobs submitted to threadpool when peerWorker is unavailable.

Comment (zaynt4606) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:620]: `fileWriter.decrementPendingWrites()` here is to make up the operation in `writeLocalData` within the `Runnable` .

**Pull request closed without merging**